src/services/document-service.js

```diff
@@ -81,35 +86,52 @@ class DocumentService {
   async getSearchableDocumentsMetadataByTags(searchQuery) {
     const tokens = searchQuery.trim().split(/\s+/);
 
-    const positiveTokens = new Set(tokens
-      .filter(token => !token.startsWith('-'))
-      .filter(token => token.length > 2)
-      .map(token => token.toLowerCase()));
-
-    const negativeTokens = new Set(tokens
-      .filter(token => token.startsWith('-'))
-      .map(token => token.substr(1))
-      .filter(token => token.length > 2)
-      .map(token => token.toLowerCase()));
+    const allPositiveTokens = new Set();
+    const positiveExactTokens = new Set();
+    const positivePartialTokens = new Set();
+    const allNegativeTokens = new Set();
+
+    for (const token of tokens) {
+      const isNegative = token.length > 1 && token.startsWith('-');
+      const rawToken = token.substr(isNegative ? 1 : 0).toLowerCase();
+      if (isNegative) {
+        allNegativeTokens.add(rawToken);
+      } else if (rawToken.length < DOCUMENT_PARTIAL_SEARCH_THRESHOLD) {
+        allPositiveTokens.add(rawToken);
+        positiveExactTokens.add(rawToken);
+      } else {
+        allPositiveTokens.add(rawToken);
+        positivePartialTokens.add(rawToken);
+      }
+    }
 
-    if (!positiveTokens.size) {
+    if (!positiveExactTokens.size && !positivePartialTokens.size) {
```

Comment:
Basically `!allPositiveTokens.size`, right?

Reply:
Yes, you're right (I introduced the `allPositiveTokens` at a later point not realizing that it can also simplify this condition), changed!